hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/volume/TestVolumeHealthMetrics.java

```diff
@@ -0,0 +1,167 @@
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.common.volume;
+
+import static org.apache.ozone.test.MetricsAsserts.assertGauge;
+import static org.apache.ozone.test.MetricsAsserts.getMetrics;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.hadoop.metrics2.MetricsRecordBuilder;
+import org.apache.hadoop.metrics2.MetricsSystem;
+import org.apache.hadoop.metrics2.lib.DefaultMetricsSystem;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Test for VolumeHealthMetrics.
+ */
+class TestVolumeHealthMetrics {
+
+  private MetricsSystem metrics;
+  private VolumeSet mockVolumeSet;
+  private List<StorageVolume> volumes;
+  private VolumeHealthMetrics volumeHealthMetrics;
+
+  @BeforeEach
+  void setUp() {
+    metrics = DefaultMetricsSystem.instance();
+    metrics.init("test");
+
+    mockVolumeSet = mock(VolumeSet.class);
+    volumes = new ArrayList<>();
+
+    volumeHealthMetrics = VolumeHealthMetrics.create(StorageVolume.VolumeType.DATA_VOLUME, mockVolumeSet);
+  }
+
+  @AfterEach
+  void tearDown() {
+    if (volumeHealthMetrics != null) {
+      volumeHealthMetrics.unregister();
+    }
+
+    metrics.stop();
+    metrics.shutdown();
+  }
+
+  private StorageVolume createMockVolume(StorageVolume.VolumeState state) {
+    StorageVolume volume = mock(StorageVolume.class);
+    when(volume.getStorageState()).thenReturn(state);
+    return volume;
+  }
+
+  /**
+   * Test metrics with empty volume list.
+   */
+  @Test
+  void testEmptyVolumeList() {
+    when(mockVolumeSet.getVolumesList()).thenReturn(new ArrayList<>());
+
+    MetricsRecordBuilder metricsRecords = getMetrics(volumeHealthMetrics);
+
+    assertGauge("TotalVolumes", 0, metricsRecords);
+    assertGauge("NumHealthyVolumes", 0, metricsRecords);
+    assertGauge("NumFailedVolumes", 0, metricsRecords);
+  }
+  
+  /**
+   * Test that metrics are correctly initialized and reported.
+   */
+  @Test
+  void testBasicMetrics() {
+    StorageVolume normalVolume1 = createMockVolume(StorageVolume.VolumeState.NORMAL);
+    StorageVolume normalVolume2 = createMockVolume(StorageVolume.VolumeState.NORMAL);
+    StorageVolume failedVolume = createMockVolume(StorageVolume.VolumeState.FAILED);
+
+    volumes.add(normalVolume1);
+    volumes.add(normalVolume2);
+    volumes.add(failedVolume);
+
+    when(mockVolumeSet.getVolumesList()).thenReturn(volumes);
```

Comment:
It would be better to test on a real `VolumeSet` instance. The mock here fails to catch the bug where failed volumes are not reported. `TestVolumeSet` should provide enough scaffolding for us to add tests there, and a `@VisibleForTesting` method to get the metrics from the `MutableVolumeSet` and check them in the test can be added. Alternatively the metrics instance may be able to be retrieved using its registered name only instead of a getter.